wordpress-mobile/WordPress-Android
# [Dependency Updates] Update `androidxLifecycleVersion` to 2.5.1

Parent #17561
Batch Branch: [deps/main-batch-androidx-core](https://github.com/wordpress-mobile/WordPress-Android/tree/deps/main-batch-androidx-core)

🚫 Blocked By:
- [x] #17560 and its `kotlinxCoroutinesVersion` update to `1.6.4` (for more info see 0ead5d34e0433a24f4e82e9d04ab2366359dd146)

This PR updates `androidxLifecycleVersion` to [2.5.1](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.5.1).

Also, as part of this update the below transitive dependencies were added:
- On the `Image Editor` module (2d936392992e2e265e1e412a069add1e33907f7a):
    - `androidx.lifecycle:lifecycle-common`
    - `androidx.lifecycle:lifecycle-runtime`
    - `androidx.lifecycle:lifecycle-viewmodel-savedstate`
- On the `Editor` module (482f47be106955bb3ca66b691949b7a97521db57):
    - `androidx.lifecycle:lifecycle-common`
    - `androidx.lifecycle:lifecycle-runtime`
    - `androidx.lifecycle:lifecycle-livedata-core`
- On the `WordPress` module (3cc47ab3fee8ef9586aecfb17cf4ac1f9d5b0f43):
    - `androidx.lifecycle:lifecycle-common`
    - `androidx.lifecycle:lifecycle-runtime`
    - `androidx.lifecycle:lifecycle-viewmodel`
    - `androidx.lifecycle:lifecycle-livedata-core`
    - `androidx.lifecycle:lifecycle-livedata`

FYI: @oguzkocer I am mentioning you here so that you are aware that progress on this dependency update, and as such, the whole #17558 parent issue is being paused as #17560 and its `kotlinxCoroutinesVersion` update to `1.6.4` is now blocking.

-----

In addition to the dependency update commits, a couple of additional test related commits were added to fix a couple of failing tests suites, see below:

1. [Fix conversation notifications view model test.](https://github.com/wordpress-mobile/WordPress-Android/pull/17643/commits/a0537cb69c0ee4778e2cf9abda46b7f77cb91e1e)
2. [Fix reader discover view model test.](https://github.com/wordpress-mobile/WordPress-Android/pull/17643/commits/fec242045364e9a067904077f0de2b6c71571849)

However, I am still not sure why those tests were passing when `AndroidX Lifecycle` is on `2.4.1`, and only started failing after updating the `AndroidX Lifecycle` library to `2.5.1`. They should have been failing with `2.4.1` as well. My assumption is that the update to `2.5.1` made the library a bit more strict in terms of testing, some bugs related to that got fixed, and thus, things that were previously falsely working are now failing unless the setup is concrete. 🤔

**UPDATE**: I actually found why this started happening, and it is indeed a bug fix, which actually improves our testing. Thus, this is nothing concerning in terms of those tests starting failing due to a potential change that might be affecting our main sources. See [2.5.0-rc01](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.5.0-rc01) release notes (`Bug Fixes`). Copy-pasted below for ease of reference:

`MediatorLiveData.addSource()` now throws a `NullPointerException` when passed a `null` source instead of
propagating the `null` source to observers.([Ibd0fb](https://android-review.googlesource.com/#/q/Ibd0fbe4d6efef0c4b6bab85eb27b4ed1b872382d), [b/123085232](https://issuetracker.google.com/123085232))

-----

PS: @irfano I added you as the main reviewer, that is, in addition to the @wordpress-mobile/apps-infrastructure  team itself, but randomly, since I just wanted someone from the WordPress team to be aware of and sign-off on that change for WPAndroid.

-----

To test:

1. See the dependency tree diff result and verify correctness.
3. Thoroughly smoke test both, the WordPress and Jetpack apps, and see if they both work as expected.

-----

### Merge instructions

- [x] Wait for #17560 issue to get resolved and the `kotlinxCoroutinesVersion` update to `1.6.4` merged to `trunk`.
- [x] Update PR's base `deps/main-batch-androidx-core` branch with latest `trunk`.
- [x] Update PR with latest `deps/main-batch-androidx-core`.
- [x] Prepare and update PR to a `Ready for review` state.
- [x] Assign a reviewer.
- [x] Merge PR to `deps/main-batch-androidx-core`.

-----

## Regression Notes
1. Potential unintended areas of impact

    - Potential breakage of screens and core functionality as `androidx.lifecycle` is a library that is being added as a transitive dependency across lots of `androidx` and `com.google` and `other` libraries.
    - Some of the transitive dependencies added might be causing some kind of misbehaviour.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

4. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

PR submission checklist:
